tmt/schemas/provision/hardware.yml

```diff
@@ -0,0 +1,220 @@
+
+    properties:
+      type:
+        type: string
+
+    additionalProperties: false
+
+    # enforce at least one property - we don't care which one, but we don't want
+    # empty `network`.
+    minProperties: 1
+
+  # HW requirements: `network` block
+  hw_networks:
+    type: array
+    items:
+      "$ref": "#/definitions/hw_network"
+
+  # HW requirements: `virtualization` block
+  hw_virtualization:
+    type: object
+
+    properties:
+      is-virtualized:
+        type: boolean
+
+      is-supported:
+        type: boolean
+
+      hypervisor:
+        type: string
+        enum:
+          - nitro
+          - kvm
+          - xen
+
+    additionalProperties: false
+
+    # enforce at least one property - we don't care which one, but we don't want
+    # empty `boot`.
+    minProperties: 1
+
+  # HW requirements: basic block
+  hw_block:
+    type: object
+
+    properties:
+      boot:
+        "$ref": "#/definitions/hw_boot"
+
+      cpu:
+        "$ref": "#/definitions/hw_cpu"
+
+      disk:
+        "$ref": "#/definitions/hw_disks"
+
+      hostname:
+        "$ref": "#/definitions/hw_hostname"
+
+      memory:
+        "$ref": "#/definitions/hw_memory"
+
+      network:
+        "$ref": "#/definitions/hw_networks"
+
+      virtualization:
+        "$ref": "#/definitions/hw_virtualization"
+
+    additionalProperties: false
+
+    # enforce at least one property - we don't care which one, but we don't want
+    # empty `block`.
+    minProperties: 1
+
+  # HW requirements: conditions
+  hw_and:
+    type: object
+    properties:
+      "and":
+        type: array
+        items:
+          anyOf:
+            - "$ref": "#/definitions/hw_boot"
+            - "$ref": "#/definitions/hw_cpu"
+            - "$ref": "#/definitions/hw_disks"
+            - "$ref": "#/definitions/hw_networks"
+            - "$ref": "#/definitions/hw_virtualization"
+            - "$ref": "#/definitions/hw_block"
+            - "$ref": "#/definitions/hw_and"
+            - "$ref": "#/definitions/hw_or"
+
+    required:
+      - "and"
+
+  hw_or:
+    type: object
+    properties:
+      "or":
+        type: array
+        items:
+          anyOf:
+            - "$ref": "#/definitions/hw_boot"
+            - "$ref": "#/definitions/hw_cpu"
+            - "$ref": "#/definitions/hw_disks"
+            - "$ref": "#/definitions/hw_networks"
+            - "$ref": "#/definitions/hw_virtualization"
+            - "$ref": "#/definitions/hw_block"
+            - "$ref": "#/definitions/hw_and"
+            - "$ref": "#/definitions/hw_or"
+
+    required:
+      - "or"
+
+type: object
+additionalProperties: false
+
+properties:
+  anyOf:
+    - "$ref": "#/definitions/hw_boot"
+    - "$ref": "#/definitions/hw_cpu"
+    - "$ref": "#/definitions/hw_disks"
+    - "$ref": "#/definitions/hw_networks"
+    - "$ref": "#/definitions/hw_virtualization"
+    - "$ref": "#/definitions/hw_block"
+    - "$ref": "#/definitions/hw_and"
+    - "$ref": "#/definitions/hw_or"
```

Comment:
The `memory` key is missing here. Tested with a simple plan example:

```yaml
provision:
    how: artemis
    hardware:
        memory: 8GB
```

Aren't the key names missing here? Shouldn't the definition here be similar as in `hw_block` above?

Reply:
I'm not sure, it looks correct, but I am too seeing validation errors with your example. I don't know why, yet, but the schema should work, it's been used by Artemis for months. Maybe there's a typo or a copy/paste error.

Reply:
With this patch to hardware schema, things start to work:

```patch
diff --git a/tmt/schemas/provision/artemis.yaml b/tmt/schemas/provision/artemis.yaml
index b7c8397..6fabcbb 100644
--- a/tmt/schemas/provision/artemis.yaml
+++ b/tmt/schemas/provision/artemis.yaml
@@ -51,7 +51,7 @@ properties:
     type: string
 
   hardware:
-    $ref: "/schemas/provision/hardware#"
+    $ref: "/schemas/provision/hardware#/definitions/hardware"
 
   pool:
     type: string
diff --git a/tmt/schemas/provision/hardware.yaml b/tmt/schemas/provision/hardware.yaml
index be79a7a..b4a58f4 100644
--- a/tmt/schemas/provision/hardware.yaml
+++ b/tmt/schemas/provision/hardware.yaml
@@ -203,18 +203,13 @@ definitions:
     required:
       - "or"
 
-type: object
-additionalProperties: false
-
-properties:
-  anyOf:
-    - "$ref": "#/definitions/hw_boot"
-    - "$ref": "#/definitions/hw_cpu"
-    - "$ref": "#/definitions/hw_disks"
-    - "$ref": "#/definitions/hw_networks"
-    - "$ref": "#/definitions/hw_virtualization"
-    - "$ref": "#/definitions/hw_block"
-    - "$ref": "#/definitions/hw_and"
-    - "$ref": "#/definitions/hw_or"
-
-  required: []
+  hardware:
+    anyOf:
+      - "$ref": "#/definitions/hw_boot"
+      - "$ref": "#/definitions/hw_cpu"
+      - "$ref": "#/definitions/hw_disks"
+      - "$ref": "#/definitions/hw_networks"
+      - "$ref": "#/definitions/hw_virtualization"
+      - "$ref": "#/definitions/hw_block"
+      - "$ref": "#/definitions/hw_and"
+      - "$ref": "#/definitions/hw_or"
```

1. add a definition of the whole `hardware` key as a definition, not an object described by the schema in `hardware.yaml`
2. this definition is not yet another object but uses `anyOf` directly, it "becomes" one of the listed schemas as needed.

The schema in this PR was probably introducing one extra level of objects by putting this `anyOf` under `properties` in `hardware.yaml`, and this was the issue.

Reply:
@happz thanks for the investigation and solution, I applied it.